### PR TITLE
[DD-555] add mailerq rest api client

### PR DESCRIPTION
**Pull request by elsupergomez**
This PR is for add a _dotnet_ client for [MailerQ Rest Api](https://www.mailerq.com/documentation/5.13/rest-api-v1)

Inspired in #33 

- [x] Allow `CancellationToken` parameter

### CODE REVIEW

File: MailerQ/RestApi/Client/IRestApiClient.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        Task<HttpStatusCode> Delete<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;
+
+        /// <summary>
+        /// Request an HTTP Post create or update an instance of the model
+        /// </summary>
+        /// <typeparam name="T">Model type that accept the endpoint</typeparam>
+        /// <param name="model">Object model to send in the request body</param>
+        /// <param name="cancellationToken">The request cancellationToken</param>
+        /// <returns>The HTTP status code returned on the request</returns>
+        Task<HttpStatusCode> Post<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;
```
Comment (ms-darianbenito) [inline MailerQ/RestApi/Client/IRestApiClient.cs:37]: What do you think about something like this?
```suggestion
        /// <summary>
        /// Request an HTTP Get to obtain a collection of instances of the model
        /// </summary>
        /// <typeparam name="T">Model type that return the endpoint</typeparam>
        /// <param name="cancellationToken">The request cancellationToken</param>
        /// <returns>Collection of objects returned by the endpoint. Null if http status code of the request is different than OK</returns>
        Task<IRestApiClientResponse> Get<T>(CancellationToken cancellationToken = default) where T : IRestApiModel;

        /// <summary>
        /// Request an HTTP Delete remove an instance of the model
        /// </summary>
        /// <typeparam name="T">Model type that accept the endpoint</typeparam>
        /// <param name="model">Object model to send in the request body</param>
        /// <param name="cancellationToken">The request cancellationToken</param>
        /// <returns>The HTTP status code returned on the request</returns>
        Task<IRestApiClientResponse> Delete<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;

        /// <summary>
        /// Request an HTTP Post create or update an instance of the model
        /// </summary>
        /// <typeparam name="T">Model type that accept the endpoint</typeparam>
        /// <param name="model">Object model to send in the request body</param>
        /// <param name="cancellationToken">The request cancellationToken</param>
        /// <returns>The HTTP status code returned on the request</returns>
        Task<IRestApiClientResponse> Post<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;
```

```
>  public interface IRestApiClientResponse
>     {
>         HttpStatusCode HttpStatusCode { get; }
>     }
>     
>     public class BasicRestApiClientResponse : IRestApiClientResponse
>     {
>         public HttpStatusCode HttpStatusCode { get; }
> 
>         public BasicRestApiClientResponse(HttpStatusCode httpStatusCode)
>         {
>             HttpStatusCode = httpStatusCode;
>         }
>     }
>     
>     public class DataRestApiClientResponse<T> : BasicRestApiClientResponse
>     {
>         public T Data { get; }
> 
>         public DataRestApiClientResponse(HttpStatusCode httpStatusCode, T data) : base(httpStatusCode)
>         {
>             Data = data;
>         }
>     }
```

Comment (ms-darianbenito) [reply MailerQ/RestApi/Client/IRestApiClient.cs]: We made part of this refactor by inserting `!fixup` in existing commits.

File: MailerQ/RestApi/Client/IRestApiClientResponse.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+using System.Net;
+
+namespace MailerQ.RestApi.Client
+{
+    public interface IRestApiClientResponse
```
Comment (elsupergomez) [inline MailerQ/RestApi/Client/IRestApiClientResponse.cs:6]: @ms-darianbenito Could you add the summary documentation?

File: MailerQ/RestApi/Client/IRestApiClient.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        /// <returns>The HTTP status code returned on the request</returns>
+        Task<IRestApiClientResponse> Delete<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;
+
+        /// <summary>
+        /// Request an HTTP Post create or update an instance of the model
+        /// </summary>
+        /// <typeparam name="T">Model type that accept the endpoint</typeparam>
+        /// <param name="model">Object model to send in the request body</param>
+        /// <param name="cancellationToken">The request cancellationToken</param>
+        /// <returns>The HTTP status code returned on the request</returns>
```
Comment (elsupergomez) [inline MailerQ/RestApi/Client/IRestApiClient.cs:36]: @ms-darianbenito please change the return description

File: MailerQ/RestApi/Client/IRestApiClientResponse.cs (hunk 2 above)
Comment (ms-darianbenito) [reply MailerQ/RestApi/Client/IRestApiClientResponse.cs:9]: Done!

File: MailerQ/RestApi/Client/IRestApiClient.cs (hunk 3 above)
Comment (ms-darianbenito) [reply MailerQ/RestApi/Client/IRestApiClient.cs]: Done! I changed all return descriptions

File: MailerQ/RestApi/IRestApiClient.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        /// <summary>
+        /// Request an HTTP Get to obtain a collection of instances of the model
+        /// </summary>
+        /// <typeparam name="T">Model type that return the endpoint</typeparam>
+        /// <param name="restApiRequestFilter">Specific definition of request the Rest API</param>
+        /// <param name="cancellationToken">The request cancellationToken</param>
+        /// <returns>Response information returned on the request including the collection of instances of the model</returns>
+        Task<IRestApiResponse<T>> Get<T>(
+            IRestApiRequest<T> restApiRequestFilter = default,
```
Comment (ms-darianbenito) [inline MailerQ/RestApi/IRestApiClient.cs:19]:
```suggestion
            IRestApiRequest<T> restApiRequest = default,
```

File: MailerQ/RestApi/IRestApiClient.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    /// <summary>
+    /// Contract that represent the HTTP methods availables in the MailerQ Rest API
+    /// </summary>
+    public interface IRestApiClient
+    {
+        /// <summary>
+        /// Request an HTTP Get to obtain a collection of instances of the model
+        /// </summary>
+        /// <typeparam name="T">Model type that return the endpoint</typeparam>
+        /// <param name="restApiRequestFilter">Specific definition of request the Rest API</param>
```
Comment (ms-darianbenito) [inline MailerQ/RestApi/IRestApiClient.cs:15]:
```suggestion
        /// <param name="restApiRequest">Specific definition of request the Rest API</param>
```

File: MailerQ/RestApi/IRestApiRequest.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    /// </summary>
+    /// <typeparam name="T">Model type that accept the endpoint</typeparam>
+    public interface IRestApiRequest<T> where T : IRestApiModel
+    {
+        /// <summary>
+        /// The query string parameters
+        /// </summary>
+        string QueryString { get; }
+    }
+
```
Comment (ms-darianbenito) [inline MailerQ/RestApi/IRestApiRequest.cs:14]: Just a detail 
```suggestion
```

File: MailerQ/RestApi/IRestApiClient.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            CancellationToken cancellationToken = default) where T : IRestApiModel;
+
+        /// <summary>
+        /// Request an HTTP Post create or update an instance of the model
+        /// </summary>
+        /// <typeparam name="T">Model type that accept the endpoint</typeparam>
+        /// <param name="model">Object model to send in the request body</param>
+        /// <param name="cancellationToken">The request cancellationToken</param>
+        /// <returns>Response information returned on the request</returns>
+        Task<IRestApiResponse<T>> Post<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;
```
Comment (ms-darianbenito) [inline MailerQ/RestApi/IRestApiClient.cs:29]: What do you think about to use `Async` suffix like some [IQueuePublisher](https://github.com/MakingSense/mailerq-dotnet-sdk/blob/19830b85de0a15fb6509259736ea96ec317b0168/MailerQ/Core/IQueuePublisher.cs) contract´s methods?
```suggestion
        Task<IRestApiResponse<T>> PostAsync<T>(T model, CancellationToken cancellationToken = default) where T : IRestApiModel;
```

Comment (elsupergomez) [reply MailerQ/RestApi/IRestApiClient.cs:29]: I see that too. Thanks!

File: MailerQ.Test/RestApiClientTest.cs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+            yield return new object[] { specimens.Create<PoolIP>(), $"{Version}/poolips" };
+            yield return new object[] { specimens.Create<IpAddress>(), $"{Version}/poolips" };
+            yield return new object[] { specimens.Create<Suppression>(), $"{Version}/suppressions" };
+            yield return new object[] { specimens.Create<ExternalMTA>(), $"{Version}/externalmtas" };
+            yield return new object[] { specimens.Create<ExternalIp>(), $"{Version}/externalmtas" };
+        }
+
+        [Theory]
+        [MemberData(nameof(RestApiModelEndpointSpecimens))]
+        public async Task Should_request_to_the_expected_endpoint_name_for_model_type<T>(T model, string endpoint) where T : IRestApiModel
```
Comment (ms-darianbenito) [inline MailerQ.Test/RestApiClientTest.cs:96]: Great!

**Pull request merged**